### ISSUE TITLE
Kubectl edit should use replace instead of patch

### ISSUE THREAD

**Issue by pwittrock** [kubernetes/kubernetes#35016]
Currently, when using kubectl edit, the end user is presented with the full yaml, makes some changes after looking at it, and then patches the existing yaml.  Just doing a replace and throwing a user facing error on a optimistic lock failure seems like a better approach since this is user facing and the full yaml is present.

**Comment by pwittrock** [comment]
@ymqytw If this is a simple change I'd prioritize it for sometime this week.

**Comment by kargakis** [comment]
@kubernetes/kubectl

**Comment by ymqytw** [comment]
It should be a simple change.

**Comment by ghodss** [comment]
When you say "optimistic lock failure", what mechanism exactly are you referring to?

**Comment by pwittrock** [comment]
@ghodss

@caesarxuchao  Correct me if I am wrong, but don't we get the version of the object when it is read and send it back with the updates.  The server then does a conditional write on the object version being the same as when it was read, and if it has changed the write fails (optimistic lock failure).

It was my understanding that this is how one can safely replace objects without worrying about race conditions from multiple entities trying to update the object concurrently.

**Comment by caesarxuchao** [comment]
@pwittrock you are right.

**Comment by smarterclayton** [comment]
The reason we use patch is because for any object under active management edit is unusable (editing a node is impossible, editing a deployment or RC is annoying, etc).  The current logic should be checking that the patch doesn't intersect with a mutation on the server, but we could be broaden that to be "safer".  Patch is pretty important.

**Comment by pwittrock** [comment]
@smarterclayton WDYT of having replace be the default behavior and patch be flag driven?  How often are Deployments are RCs updated frequently enough by other parties to cause a locking exception?

**Comment by pwittrock** [comment]
Thanks for clarification on editing Nodes.

We are definitely working on improving the limitations of patch.

**Comment by smarterclayton** [comment]
So imagine this scenario.  You run a create command.  Then realize you forgot to set an environment variable.  You run edit.  Make the change.  Save.  Conflict!   You edit again.  Save.  Conflict!  You hate kube and quit your day job.

When we rolled edit out originally in openshift it survived about 2 weeks and then we rewrote it with patch because it was completely frustrating to users.   Since then, no complaints.  

The patch logic should check for conflicts, but we can do better at telling users what changed when a  change happened (in the comments up top).  I've never met a user who was using edit who cared about any edit except one that overwrote their own (since if they won the race they wouldn't know anyway).

**Comment by smarterclayton** [comment]
We *could* also check for spec changes.  That's much rarer, but it doesn't apply to all types.  So we'd need something in swagger that told us whether a field was "safe" or not, or come up with a heuristic, then the heuristic would need a flag because not all of our objects have spec and status.

**Comment by smarterclayton** [comment]
A swagger field that said "user-editable" has been something we've discussed for a while - if we found one of those on a type we could have different behavior on those fields vs the others, but we probably still have to deal with third party resources and extension resources that won't have those.

**Comment by pwittrock** [comment]
@smarterclayton 
user-editable in swagger would be helpful

The other concern I have with PATCH is that sometimes it doesn't do the right thing.   For example: #35163 #32398.

**Comment by smarterclayton** [comment]
And  kubectl apply umbrella issue tracker
<https://github.com/kubernetes/kubernetes/issues/35234> really brings it
home - we can't even trust it for that, and it's critical.  We didn't
commit resources for 1.5 but I think we need to do some fuzzing (human or
automated) that starts bringing patch / apply under control.  Also STMRG
doesn't work without the resource definition on hand, which is very
frustrating for generic transformation tools (like edit is about to become
as soon as we start supporting extensions).

**Comment by pwittrock** [comment]
>
STMRG

Sorry what is that an acronym for ? :P

Agreed that requiring the resource definition to be linked in causes all sorts of problems - such as dropping fields from version skewed clients.

@ymqytw Has been working through some apply issues in 1.5 and we expect the merging of primitive lists types to make it in this quarter (PR is out now).  Getting more resources focused on this in 1.6 would be great too.

In summary, is it correct to say that you think we need to keep `kubectl edit` using patch due to the issue with concurrent edits?

**Comment by smarterclayton** [comment]
STrategic MeRGe :)

Primarily concurrent management by controllers, and only weakly by other end users.  Most controller management is correlated to creation time, and many edits are done correlated to creation time as well.

**Comment by pwittrock** [comment]
I knew it couldn't be `Street Manager`

**Comment by kargakis** [comment]
> @smarterclayton WDYT of having replace be the default behavior and patch be flag driven? How often are Deployments are RCs updated frequently enough by other parties to cause a locking exception?

Both are updated frequently during rollouts by their respective controllers. Trying to edit something just after a rollout is pretty common.

**Comment by pwittrock** [comment]
Alright I guess we keep it as patch.

**Comment by soltysh** [comment]
@pwittrock since all your doubts have been cleared out can we close this issue in that case?
